### DISCUSSION TITLE
RespawnBsp History & Progress Reports

### DISCUSSION THREAD

**Discussion by snake-biscuits** [snake-biscuits/bsp_tool#78] [Show and tell]
Documenting the latest theories and experiments reverse engineering the RespawnBsp format(s)

**Comment by snake-biscuits** [comment]
I'm afraid it's been... 9 years...
# Timeline of Titanfall modding

## 2014
### January
Before Titanfall's retail release, but after the public betas and tech tests, [Cra0kalo](https://github.com/cra0) makes plans to create modding tools[^1]
### February
[Cra0kalo](https://github.com/cra0) reverses Respawn's `.vpk` archive format and publicly releases tools (closed source) to extract their files[^2]
### March
By early March `.bsp` files had been dug into, with strings identifying multiple lump names found in `bin/x64_retail/bsppack.dll`
[ata4](https://github.com/ata4) starts work on bringing Titanfall "support" to [bspsrc](https://github.com/ata4/bspsrc), a popular Source Engine map decompiler[^3]
### August
[Cra0kalo](https://github.com/cra0) theorises (incorrectly) that the map format hasn't changed much since Portal 2[^4]

## 2016
Titanfall 2 releases

## 2017
### May
The Titanfall 2 single-player only [Icepick](https://titanfallmods.com/) mod launcher enters development[^5]

## 2020
### March
[snake-biscuits](https://github.com/snake-biscuits) is killed by a CS:GO hacker with a YouTube URL for a name; they follow this link, and find cra0's blog
Discovering VPK Tool, they add `mods/titanfall2.py` to [bsp_tool](https://github.com/snake-biscuits/bsp_tool)[^6]

Various projects to create modded clients for Titanfall: Online, Titanfall 2 & Apex Legends cross paths across multiple Discord servers
[snake-biscuits](https://github.com/snake-biscuits) gets involved, trying to port maps between all 3 games

## 2021
### August
The Apex Legends mod client [R5Reloaded](https://r5reloaded.com/) releases to the public[^7]
### December
[BobTheBob](https://github.com/BobTheBob9) releases `v1.0.0` of the Titanfall 2 Northstar client publically, just in time for Christmas[^8]

## 2022
### April
[F1F7Y](https://github.com/F1F7Y) begins work on [MRVN-Radiant](https://github.com/MRVN-Radiant/MRVN-Radiant), a fork of the [NetRadiant-custom](https://github.com/Garux/netradiant-custom) level editing software[^9]

### June
[barnabwhy](https://github.com/barnabwhy) releases their open source VPK tool: [TFVPKTool](https://github.com/barnabwhy/TFVPKTool)[^10] (Cra0's tool was closed source)
This forms the backend for [HarmonyVPKTool](https://github.com/harmonytf/HarmonyVPKTool)

By the end of the year [MRVN-Radiant](https://github.com/MRVN-Radiant/MRVN-Radiant) is compiling basic maps for all 3 games
Momentum researching the `.bsp` format continues to grow exponentially, with better and better tools helping us to experiment & learn

## 2023
Team Floppa Online real?

[^1]: 2014, Jan 3 - Twitter: [Cra0kalo vows to create modding tools](https://twitter.com/cra0kalo/status/418991750225465344)
[^2]: 2014, Feb 15 - Cra0 Blog: [Cra0kalo releases VPK Tool](https://cra0.net/blog/posts/archived/2014/titanfall-vpk-tool/)
[^3]: 2014, Mar 4 - GitHub (ata4/bspsrc): [working begins on Titanfall support](https://github.com/ata4/bspsrc/commit/0eb6cc7f4d7713acf7e9e1f0e92cc5072cd7f087)
[^4]: 2014, Aug 3 - Cra0 Blog: [Cra0kalo talks about `.bsp` & xbox360 `.bpk`](https://cra0.net/blog/posts/archived/2014/titanfall-vpktool-3-3-update/)
[^5]: 2017, May 10 - GitHub (Titanfall-Mods/Titanfall-2-Icepick) [Initial Commit](https://github.com/Titanfall-Mods/Titanfall-2-Icepick/tree/a540bdd2aded4d9021843736cec69e66f69c7ef6)
[^6]: 2020, Mar 1 - GitHub (snake-biscuits/bsp_tool): [research begins on Titanfall 2](https://github.com/snake-biscuits/bsp_tool/blob/a9f6749d59a5575cc575bdb18f243ab18dec6a20/mods/titanfall2.py)
[^7]: 2021, Aug 14 - GitHub (Mauler125/r5sdk): [R5Reloaded v1.4.0](https://github.com/Mauler125/r5sdk/releases/tag/v1.4)
[^8]: 2021, Dec 24 - GitHub (R2Northstar/Northstar): [Northstar v1.0.0](https://github.com/R2Northstar/Northstar/releases/tag/v1.0.0)
[^9]: 2022, Apr 22 - GitHub (MRVN-Radiant/MRVN-Radiant): [Initial Commit](https://github.com/MRVN-Radiant/MRVN-Radiant/tree/8b5f5cfd6da13a27367d238b05124e47be0b69b3)
[^10]: 2022, Jun 30 - GitHub (barnabwhy/TFVPKTool): [Initial Commit](https://github.com/barnabwhy/TFVPKTool/commit/fa9c1829ab592ca16ff9dc3b3795e8e1a14bdd44)

**Reply by snake-biscuits** [reply]
>2017 - May: The Titanfall 2 single-player only [Icepick](https://titanfallmods.com/) mod launcher enters development

Icepick is significant because it served as a base / inspiration for Northstar
The Icepick devs also managed to create a python script for converting Titanfall 2 `.bsp` files into `.obj` models
Likely reversed from a function present in Titanfall 2 itself, this code formed the bedrock of `bsp_tool` research into Titanfall 2 maps

**Reply by snake-biscuits** [reply]
Preserving some screencaps of citations just in case Twitter totally collapses
![cra0_declares_war](https://github.com/snake-biscuits/bsp_tool/assets/36507175/fed85f06-6602-42c9-a1f0-ad36cebfaf45)
![cra0_vpk_tool_release](https://github.com/snake-biscuits/bsp_tool/assets/36507175/b02150b7-5461-49c5-a3e9-b60cf04b7d94)
![cra0_vpk_tool_3 4_plans](https://github.com/snake-biscuits/bsp_tool/assets/36507175/e415ef51-9cc4-459c-b4dd-cc4f0d0e6a24)

**Comment by snake-biscuits** [comment]
In 2023, we have a pretty firm grasp on how Respawn Source renders `.bsp` files
We're really starting to dig into the Visibility (culling, occlusion, depth sorting etc.) systems, as well as physics

Map decompilers are good enough to be educational, and we have some understanding of how to get lighting in-game

A lot of our development has been focused on Titanfall 2, but we can test maps on all 3 games (including the currently private TF:O client)
We've ported TF:O's mp_box to Titanfall 2 & helped the R5Reloaded team create World's Edge: After Dark

There's a lot of big ideas ahead, but we're still slowly building up the tools
Very happy to be at the point where progress is something that can be expressed in screenshots and videos
[mp_firing_range_inbounds_grenade_crash.webm](https://user-images.githubusercontent.com/36507175/212232825-425dfd26-9afb-47f0-a5cb-99a0d0152b82.webm)
![x_marks_the_spawn](https://user-images.githubusercontent.com/36507175/212232915-b90beb20-c735-4c94-ad60-7ae787986a9c.png)

**Comment by snake-biscuits** [comment]
Currently I'm working on mimicking the way meshes are sorted in official maps (transparency, decals, & skybox mesh groups)
This is happening in [my fork of MRVN-Radiant](https://github.com/snake-biscuits/MRVN-Radiant)

I also want to document & add support for all the various entities available to mappers, over at [MRVN-entities](https://github.com/MRVN-Radiant/MRVN-entities)
As that comes along I'll be developing various sample maps to demonstrate how to use them for new mappers in [MRVN-Resource-Pack](https://github.com/MRVN-Radiant/MRVN-Resource-Pack)

We recently got a bit of a breakthrough with [rexx](https://github.com/r-ex) finding a way to get the member names of various `.bsp` structs
Still haven't fully explored the new information this brings, like how different lumps index one another

Meanwhile collision research has gotten so fine we're reverse engineering bitfields & flags

**Reply by snake-biscuits** [reply]
Mesh sorting results are unclear atm.
I broke ObjReferences indices (VIS)
Titanfall 2 has no 2D skybox images
sky_camera appears to do nothing
TF:O maps don’t cache materials for some reason

**Comment by snake-biscuits** [comment]
In other news, "coverage" (% of lumps understood, by number of attributes called unknown) of RespawnBsp is very high
Titanfall is now the 3rd best mapped branch script, just below Quake 1 & 3
Titanfall 2 is 5th, just ahead of GoldSrc
Apex Legends in at around number 23, just behind most Source Engine variants
Somehow Quake 2 is behind Apex, even though Q2's source code has be public for ~25 years

| Developer | Branch | Coverage |
|:-|:-|:-|
| id_software | quake3 |100.00 |
| id_software | quake | 97.67 |
| respawn| titanfall | 92.92 |
| id_software | qfusion | 88.89 |
| respawn | titanfall2 | 87.27 |
| valve | goldsrc | 86.00 |
| valve | orange_box | 75.34 |
| valve | source | 73.28 |
| respawn | apex_legends | 69.28 |
| id_software | quake2 | 67.89 |

(taken from [`docs/supported/coverage.sc`](https://github.com/snake-biscuits/bsp_tool/blob/master/docs/supported/coverage.sc))

**Reply by snake-biscuits** [reply]
The coverage percentage isn't the best way of conveying how far we've come imo, but it does the job.
Kinda funny that the *mostly* open source Quake, GoldSrc & Source Engines aren't 100% mapped yet

**Reply by snake-biscuits** [reply]
Still, 70% and above is really good. almost all branch scripts below are obscure w/ either wierd 3rd party tools or no tools at all
Though everything below 50% has official tools (if you count all 3 mohaa branch scripts as the same game)

CSO2 is in the unenviable spot of 50% support with very limited tools, and not enough time to really work on it
And Vindictus is at ~71%, but I have yet to find a client to test custom maps (though one may exist)

**Comment by snake-biscuits** [comment]
## Physics Reversing
 * #81
 * #106
 * #111

**Reply by snake-biscuits** [reply]
#81 in particular documents methodology pretty well (theory crafting & bulk testing)

See the wiki page on [spelunking](https://github.com/snake-biscuits/bsp_tool/wiki/.bsp-spelunker's-guide:-testing-a-theory) for more notes on research methodology

**Comment by snake-biscuits** [comment]
## Lighting Reversing
 * #122 
 * #176
 * #89

**Comment by snake-biscuits** [comment]
Got a big burst of reversing work done on WaterBodies recently
 * #166 
 
**tl;dr** it's the lumps involved in rendering the ocean on Habitat & Storm Point
